### PR TITLE
Add http.server.response.size metric to ASGI implementation.

### LINKED ISSUE #1782: response_size_histogram for ASGI instrumention
**Describe the solution you'd like**
Can we have `response_size_histogram` inside the `opentelemetry-instrumentation-asgi` logic?
Currently, we do have access to `content-length` header from the response inside the `OpenTelemetryMiddleware`, so it would be beneficial that we can have the ability to record response_size like duration.

**Describe alternatives you've considered**
Currently, I'm thinking about disable the FastAPI auto instrumentation and implement both response size and duration.

**Additional context**
The only instrumentator I could find that have the ability to record response_size is `opentelemetry-instrumentation-tornado` (`HTTP_SERVER_RESPONSE_SIZE`).

Let me know if that makes sense.
Thanks.

### PR DESCRIPTION
**Pull request by imanshafiei540**
Add new unit tests.

# Description

This PR adds `http.server.response.size` to ASGI instrumentation logic.

It follows naming/conventions from here: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/http-metrics.md#metric-httpserverresponsesize

It's basically using `content-length` from the response headers to record response size.

Fixes #1782 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Changed the current unit tests to support `content-length` and test `http.server.response.size`.
- [x] Tested locally with one API service built on top of the fastAPI and everything worked as expected as I was able to see old/new metrics in the Grafana dashboard.

# Does This PR Require a Core Repo Change?

- [ ] Yes. - Link to PR: 
- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added

### CODE REVIEW

**Comment by imanshafiei540** [comment]
@shalevr I’ve added new metrics and also new unit tests. Also, updated the changelog and fixed linting issues. I’m not sure, but I think we don’t need any documentation updates for my changes.
Let me know if I’m missing something or If I can refactor anything in my logic.
Thank you.

File: instrumentation/opentelemetry-instrumentation-asgi/tests/test_asgi_middleware.py [hunk 1]
```diff
@@ -602,13 +615,12 @@ async def target_asgi(scope, receive, send):
...
         self.seed_app(app)
         self.send_default_request()
-
         metrics_list = self.memory_metrics_reader.get_metrics_data()
         assertions = 0
         for resource_metric in metrics_list.resource_metrics:
             for scope_metrics in resource_metric.scope_metrics:
                 for metric in scope_metrics.metrics:
-                    if metric.name != "http.server.duration":
+                    if metric.name == "http.server.active_requests":
```
Comment (shalevr) [inline instrumentation/opentelemetry-instrumentation-asgi/tests/test_asgi_middleware.py:623]: Why is this change necessary?

Comment (imanshafiei540) [reply instrumentation/opentelemetry-instrumentation-asgi/tests/test_asgi_middleware.py:623]: It can also be:
```python
if metric.name != "http.server.duration" or metric.name != "http.server.response.size":
```
I thought this line `if metric.name == "http.server.active_requests":` would be more readable since we only have these three metrics in the expected metrics variable. Let me know if I should change it. Thanks.

Comment (shalevr) [reply instrumentation/opentelemetry-instrumentation-asgi/tests/test_asgi_middleware.py:623]: I got it, thanks

File: instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py [hunk 2]
```diff
@@ -615,6 +620,18 @@ def _get_otel_send(
     ):
         @wraps(send)
         async def otel_send(message):
+            if message.get("headers"):
```
Comment (shalevr) [inline instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:623]: Can you move this logic to a new function?
This logic is not really part of otel_send logic, so I think it will be better to keep the function simple

Comment (imanshafiei540) [reply instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py]: I refactored the implementation, and now the main logic belongs to the correct place. Thanks for your comment.

**Comment by imanshafiei540** [comment]
@shalevr 
I need to fix three tests that are failing in these packages:
1. FastAPI
2. Starlette
3. AIOHttpClient (Which I'm not sure about the issue yet)

Btw, I'm not able to use `tox` locally anymore:

`ERROR: Could not find a version that satisfies the requirement opentelemetry-semantic-conventions==0.39b0.dev (from opentelemetry-instrumentation-asgi[test]) (from versions: 0.20b0, 0.21b0, 0.22b0, 0.23b0, 0.23b2, 0.24b0, 0.25b0, 0.25b1, 0.25b2, 0.26b0, 0.26b1, 0.27b0, 0.28b0, 0.28b1, 0.29b0, 0.30b0, 0.30b1, 0.31b0, 0.32b0, 0.33b0, 0.34b0, 0.35b0, 0.36b0, 0.37b0, 0.38b0)
ERROR: No matching distribution found for opentelemetry-semantic-conventions==0.39b0.dev`

I think it's because of the recent update in the python core package, but I couldn't find a workaround for how to fix it. Could you please help me with that?

Thanks.

**Comment by shalevr** [comment]
> @shalevr I need to fix three tests that are failing in these packages:
> 
> 1. FastAPI
> 2. Starlette
> 3. AIOHttpClient (Which I'm not sure about the issue yet)
> 
> Btw, I'm not able to use `tox` locally anymore:
> 
> `ERROR: Could not find a version that satisfies the requirement opentelemetry-semantic-conventions==0.39b0.dev (from opentelemetry-instrumentation-asgi[test]) (from versions: 0.20b0, 0.21b0, 0.22b0, 0.23b0, 0.23b2, 0.24b0, 0.25b0, 0.25b1, 0.25b2, 0.26b0, 0.26b1, 0.27b0, 0.28b0, 0.28b1, 0.29b0, 0.30b0, 0.30b1, 0.31b0, 0.32b0, 0.33b0, 0.34b0, 0.35b0, 0.36b0, 0.37b0, 0.38b0) ERROR: No matching distribution found for opentelemetry-semantic-conventions==0.39b0.dev`
> 
> I think it's because of the recent update in the python core package, but I couldn't find a workaround for how to fix it. Could you please help me with that?
> 
> Thanks.

You can ignore the aiohttp-client. It's a known issue that I'm working to fix.
And the tox failed for asgi I'm not sure why, but you can run tox just for fastapi and Starlette:
tox -e test-instrumentation-fastapi
tox -e test-instrumentation-starlette

**Comment by imanshafiei540** [comment]
I've fixed unit tests for FastAPI and Starlette.

File: CHANGELOG.md [hunk 3]
```diff
@@ -29,6 +29,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   ([#1407](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/1407))
 - `opentelemetry-instrumentation-logging` Add `otelTraceSampled` to instrumetation-logging
   ([#1773](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/1773))
+- `opentelemetry-instrumentation-asgi` Add `http.server.response.size` metric
```
Comment (shalevr) [inline CHANGELOG.md:32]: Please move this entry to the Unreleased section

Comment (imanshafiei540) [reply CHANGELOG.md]: Sorry about that. Now it's updated. Thanks.

**Comment by shalevr** [comment]
Please fix the lint errors

**Comment by shalevr** [comment]
Do we also need http.server.request.size?

As described here:
https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/http-metrics.md#metric-httpserverrequestsize

**Comment by imanshafiei540** [comment]
> Do we also need http.server.request.size?
> 
> As described here: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/http-metrics.md#metric-httpserverrequestsize

@shalevr Thanks for reviewing.
I think we can implement this in a different PR (I can work on this too).
I don't see that documentation saying these two metrics are dependent on each other, and we cannot have only one of those.
Let me know what you think.

**Comment by shalevr** [comment]
> > Do we also need http.server.request.size?
> > As described here: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/http-metrics.md#metric-httpserverrequestsize
> 
> @shalevr Thanks for reviewing. I think we can implement this in a different PR (I can work on this too). I don't see that documentation saying these two metrics are dependent on each other, and we cannot have only one of those. Let me know what you think.

Sounds ok, I opened an issue for this #1865

File: instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py [hunk 4]
```diff
@@ -660,6 +670,19 @@ async def otel_send(message):
                         setter=asgi_setter,
                     )
 
+                if message.get("headers"):
```
Comment (shalevr) [inline instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:673]: I'm wondering if we can just use the get function of ASGIGetter here
It looks like the same logic.
asgi_getter.get("content-length")

Comment (imanshafiei540) [reply instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py]: You're right. I've replaced it with `asgi_getter`.
Thanks.

**Review by shalevr — APPROVED** [review]

**Comment by shalevr** [comment]
Thanks
